### PR TITLE
Add batch script for LV

### PR DESCRIPTION
**Pull request by aboucaud**

### CODE REVIEW

**Comment by aboucaud** [comment]
@EiffL could you please have a look at `azure_scripts/lotka_volterra_francois.py`

I converted your notebook into a script but it gives me NaNs and I have troubles figuring it out.
Maybe a different pair of eyes could spot something.

File: azure_scripts/lotka_volterra_francois.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  activation=jnp.sin
+)
+
+NF = partial(
+  ConditionalRealNVP, 
+  n_layers=args.nf_layers, 
+  bijector_fn=bijector
+)
+
+class SmoothNPE:
```
Comment (EiffL) [inline azure_scripts/lotka_volterra_francois.py:102]:
```suggestion
class SmoothNPE(hk.Module):
```

**Comment by EiffL** [comment]
Thanks @aboucaud looks great to me!

File: azure_scripts/lotka_volterra_francois.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+  return loss, new_params, new_opt_state
+
+# train
+batch_loss = []
+for step in range(args.n_steps):
+    mu, batch, score = get_batch(next(rng_seq))
+    l, params_nd, opt_state = update(params_nd, opt_state, mu, batch, score)
+    if (step % 100) == 0 :
+        print(f'Iter {step} / {args.n_steps}:', np.mean(batch_loss[50:]), scheduler(opt_state[1].count))
```
Comment (aboucaud) [inline azure_scripts/lotka_volterra_francois.py:151]: @EiffL is there a reason you compute the mean over all the steps starting at 50 and not the mean over the latest 50 or 100 ?

File: azure_scripts/lotka_volterra.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+opt_state = optimizer.init(params_nd)
+
+# define loss function and model update
+def loss_fn(params, weight, mu, batch, score):
+  log_prob, out = jax.vmap(
+    jax.value_and_grad(
+      lambda theta, x: nvp_nd.apply(params, theta.reshape([1, 4]), x.reshape([1, 10])).squeeze()
+    )
+  )(mu, batch)
+  return -jnp.mean(log_prob) + weight * jnp.mean(jnp.sum((out - score)**2, axis=1))
```
Comment (Justinezgh) [inline azure_scripts/lotka_volterra.py:120]: NaNs come from there if you comment `+ weight * jnp.mean(jnp.sum((out - score)**2, axis=1))` it works. For some reason, the NF gradient looks like this at some point image

File: azure_scripts/lotka_volterra_francois.py (hunk 2 above)
Comment (EiffL) [reply azure_scripts/lotka_volterra_francois.py:151]: oh right ^^ that's not what I meant to do. mean over the latest 50 is the correct thing

File: azure_scripts/lotka_volterra_francois.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  bijector_fn=bijector
+)
+
+class SmoothNPE(hk.Module):
+    def __call__(self, y):
+        net = lotka_volterra_y_bijector(y)
+        net = jnp.sin(hk.Linear(256, name='comp1')(net))
+        net = jnp.sin(hk.Linear(256, name='comp2')(net))
+        net = jnp.sin(hk.Linear(256, name='comp3')(net))
+        net = hk.Linear(args.n_components, name='comp4')(net)
```
Comment (Justinezgh) [inline azure_scripts/lotka_volterra_francois.py:108]: do we want to feed the conditional part of the NF with dimension equal to n_components ?

Comment (aboucaud) [reply azure_scripts/lotka_volterra_francois.py:108]: That's a pure interpretation on my behalf, made when I saw on the notebook that the output layer of the MLP was the same size as the number of components. 
But that may be a completely wrong assertion, in which case we should use a different argument value for it.

Comment (Justinezgh) [reply azure_scripts/lotka_volterra_francois.py:108]: I don't know ^^ @EiffL ?

**Comment by Justinezgh** [comment]
I made some tests to find the best architecture 
image
this seems to be the best architecture (the yellow dot):
``` python 
bijector_layers = [128] * 2

bijector = partial(
  AffineSigmoidCoupling, 
  layers=bijector_layers, 
  n_components=32, 
  activation=jnp.sin
)

NF = partial(
  ConditionalRealNVP, 
  n_layers=3, 
  bijector_fn=bijector
)

class SmoothNPE(hk.Module):
    def __call__(self, y):
        net = lotka_volterra_y_bijector(y)
        net = jnp.sin(hk.Linear(256, name='comp1')(net))
        net = jnp.sin(hk.Linear(256, name='comp2')(net))
        net = jnp.sin(hk.Linear(256, name='comp3')(net))
        net = jnp.sin(hk.Linear(256, name='comp4')(net))
        net = jnp.sin(hk.Linear(256, name='comp5')(net))
        net = hk.Linear(8, name='comp6')(net)
        nvp = NF(args.dimension)(net)
        return tfd.TransformedDistribution(nvp,
                                           tfb.Chain([tfb.Invert(lotka_volterra_theta_bijector),
                                                      tfb.Scale(10.),
                                                      tfb.Shift(-0.5)]))
```

File: azure_scripts/lotka_volterra_francois.py (hunk 4 above)
Comment (EiffL) [reply azure_scripts/lotka_volterra_francois.py:108]: No not n_conponents but typically the output of your compressor is of the same size (or same order of magnitude) as the number of dimensions of the distribution you are trying to estimate

**Comment by Justinezgh** [comment]
so the results I got with the previous architecture were not good, don't really have time to understand why ^^ so I just switched to the architecture with vmim compressor and got theses results:
image
image

**Comment by Justinezgh** [comment]
results: 
image

**Comment by aboucaud** [comment]
Any idea why the absolute values of the NLP are so different ?

**Comment by Justinezgh** [comment]
between simulations and simulations + score ?

**Comment by aboucaud** [comment]
Between old and new results

**Comment by Justinezgh** [comment]
the old one starts at 100 and the new one at 20

**Comment by EiffL** [comment]
I mean merge ^^ it's not good to let PRs hang around for too long.

**Comment by Justinezgh** [comment]
yes just need to make some updates
